src/optimizer/query_rewrite.c

```diff
@@ -3499,6 +3523,870 @@ qo_reduce_outer_joined_tables (PARSER_CONTEXT * parser, PT_NODE * spec, PT_NODE
+	    {
+	      reduce_pred_point_list = parser_append_node (pt_point (parser, curr_pred), reduce_pred_point_list);
+	    }
+	  else
+	    {
+	      /* Non-join predicates exist. */
+	      goto exit_on_fail_with_exclude;
+	    }
+	}
+    }
+
+  if (reduce_pred_point_list == NULL)
+    {
+      /* There are no join predicates. */
+      goto exit_on_fail_with_exclude;
+    }
+
+  /* The columns of join predicates must be in the primary key. */
+  for (curr_pred_point = reduce_pred_point_list; curr_pred_point != NULL; curr_pred_point = curr_pred_point->next)
+    {
+      SM_ATTRIBUTE *pk_cons_attr = NULL;
+      PT_NODE *pk_pred_attr = NULL;
+      PT_NODE *arg1 = NULL, *arg2 = NULL;
+      int i = 0;
+
+      curr_pred = curr_pred_point;
+      CAST_POINTER_TO_NODE (curr_pred);
+
+      assert (PT_NODE_IS_EXPR (curr_pred));
+      assert (PT_EXPR_OP (curr_pred) == PT_EQ);
+      assert (PT_NODE_IS_NAME (PT_EXPR_ARG1 (curr_pred)));
+      assert (PT_NODE_IS_NAME (PT_EXPR_ARG2 (curr_pred)));
+
+      arg1 = PT_EXPR_ARG1 (curr_pred);
+      arg2 = PT_EXPR_ARG2 (curr_pred);
+
+      if (PT_SPEC_ID (curr_pk_spec) == PT_NAME_SPEC_ID (arg1))
+	{
+	  pk_pred_attr = PT_EXPR_ARG1 (curr_pred);
+	}
+      else if (PT_SPEC_ID (curr_pk_spec) == PT_NAME_SPEC_ID (arg2))
+	{
+	  pk_pred_attr = PT_EXPR_ARG2 (curr_pred);
+	}
+      else
+	{
+	  /* Already checked before. */
+	  assert (false);
+	  ERROR_SET_ERROR_ONLY (ER_FAILED);
+	  goto cleanup_on_fail;
+	}
+
+      for (i = 0; curr_pk_cons->attributes[i] != NULL; i++)
+	{
+	  if (intl_identifier_casecmp (curr_pk_cons->attributes[i]->header.name, PT_NAME_ORIGINAL (pk_pred_attr)) == 0)
+	    {
+	      /* found */
+	      pk_cons_attr = curr_pk_cons->attributes[i];
+	      break;
+	    }
+	}
+
+      if (pk_cons_attr == NULL)
+	{
+	  /* not found */
+	  goto exit_on_fail;
+	}
+    }
+
+  assert (curr_pred_point == NULL);
+
+  {
+    PT_NODE *backup_where = NULL;
+
+    memset (&info, 0, sizeof (SPEC_CNT_INFO));
+    info.spec = curr_pk_spec;
+
+    backup_where = query->info.query.q.select.where;
+    query->info.query.q.select.where = NULL;
+
+    parser_walk_tree (parser, query, qo_get_name_cnt_by_spec_without_oncond, &info, NULL, NULL);
+
+    query->info.query.q.select.where = backup_where;
+
+    if (info.my_spec_cnt >= 1)
+      {
+	goto exit_on_fail_with_exclude;
+      }
+  }
+
+  reduce_reference_info->pk_mop = curr_pk_mop;
+  reduce_reference_info->pk_cons = curr_pk_cons;
+  reduce_reference_info->reduce_pred_point_list = reduce_pred_point_list;
+
+  return true;
+
+cleanup_on_fail:
+  parser_free_tree (parser, reduce_pred_point_list);
+  reduce_pred_point_list = NULL;
+
+  goto exit_on_fail;
+
+exit_on_fail_with_exclude:
+  er_log_debug (ARG_FILE_LINE, "%s: exclude. (spec: %s)\n", __func__, pt_print_alias (parser, curr_pk_spec));
+  reduce_reference_info->exclude_pk_spec_point_list =
+    parser_append_node (pt_point (parser, curr_pk_spec), reduce_reference_info->exclude_pk_spec_point_list);
```

Comment:
exclude "return false" intentionally to fall through?

Reply:
@beyondykk
[[failthrough]] is an explicit expression and has no role.
Therefore, it continues to the next case without a return.

Reply:
It is intended.
I will add comments below.
```
/* fallthrough */
```